### ISSUE TITLE
Questions on possibility / ideas for scripting

### ISSUE THREAD

**Issue by evan25lee** [codehz/ElementZero#58]
So I've been taking a peak at the scripting API and it seems pretty good so far. Though there are a few things that I feel like would make it better and that I myself would have a lot of use for.

1. Ability to change players gamemode: For some reason, not even the scripting api built into minecraft itself is capable of this. And while I could use a command for the built in gamemodes, things like the "hidden" spectator modes can't be accessed through it.

2. Display and edit raw packets: Now I'm not sure how possible this one is. Some way of displaying packets by itself would be fine but being able to edit them as well would be great.

3. read containers contents or NBT value via a given block coordinate: Another thing the vanilla API is for some reason not capable of.

4. GUI: Forms exist on the client side of the vanilla API, but that requires a behavior pack and is just a mess. It would be better if there was a way to handle that server side.

Consider these both requests and just wondering if they are even possible. I would imagine they are possible though I'm not sure the technical side of the game.

**Comment by Talark-Systems** [comment]
"Hidden" spectator mode can be acheived by putting the player in adventure mode, give them weakness 255 and then use the ability command to give them flight.

**Comment by IvanTurgenev** [comment]
> "Hidden" spectator mode can be acheived by putting the player in adventure mode, give them weakness 255 and then use the ability command to give them flight.

education needs to be enabled?

**Comment by Talark-Systems** [comment]
Yes

**Comment by IvanTurgenev** [comment]
> Yes

doesnt affect normal gameplay on a survival server?

**Comment by Talark-Systems** [comment]
It does add a few blocks... but other than that... no

**Comment by Bluebotlaboratories** [comment]
It does add a few blocks... but other than that... no

**Comment by IvanTurgenev** [comment]
> It does add a few blocks... but other than that... no

but to creative? not normal survival?

**Comment by Talark-Systems** [comment]
Yes, I don't think they are craftable...

**Comment by evan25lee** [comment]
> "Hidden" spectator mode can be acheived by putting the player in adventure mode, give them weakness 255 and then use the ability command to give them flight.

The "hidden" spectator is capable of flying through blocks. This wouldn't really be able to do that.

**Comment by evan25lee** [comment]
Oooh also I should of asked, would a way of custom crafting recipes be possible without having to hook? The reason why I want this stuff in scripting is because C++ is bleh in my opinion and I prefer to use JS.

**Comment by DevilLord41** [comment]
Is not you can do custom recipes by Behavior Pack?

**Comment by evan25lee** [comment]
You can do it via a behavior pack but you can also do a GUI with a behavior pack. The point is to not use one.

**Comment by evan25lee** [comment]
Also I keep thinking of ideas, a function that lets us give a player an item would be great as well. While I could use /give, it doesn't let me give certain blocks like the nether reactor.

**Comment by johnbanq** [comment]
I did thought about JS scripting, but what about performance?
The official addon mechanism is asynchronous and you can only watch things happen, not manipulate them on the go.

**Comment by evan25lee** [comment]
I mean, the vanilla API is able to  change mob components and it can run commands so I would assume that wouldn't be a problem.

**Comment by ambiennt** [comment]
> "Hidden" spectator mode can be achieved by putting the player in adventure mode, give them weakness 255 and then use the ability command to give them flight.

I would like to see a "true" spectator mode mod or built-in command in which the player has no way of interacting with the environment at all, like it is in java. The flaws with the alternative suggestion you posed are that players with these effects still have the ability to pick up items, open chests/doors, and collide hitboxes with other players. Also, with this method players still have collision with blocks. They can walk on them which would cause the game to play walking sounds. I can see how this could be a problem if a server owner was looking for a fullproof spectator mode.

**Comment by johnbanq** [comment]
But all the scripts are running asynchronously to the core game logic, so if the scripts are slow then the game still remains unaffected.
Adding support for packet modification / cancellable events would require parts of the script to be ran inside the rather critical parts of the game. That might cause major slowdown......

**Comment by SacriGrape** [comment]
> But all the scripts are running asynchronously to the core game logic, so if the scripts are slow then the game still remains unaffected.
> Adding support for packet modification / cancellable events would require parts of the script to be ran inside the rather critical parts of the game. That might cause major slowdown......

Yeah, packet modifications would probably do that. Though, at least being able to read them would be nice. Maybe force us to specify an IP we want to record from. Though, this is just a request thing. My only reason for doing it would be for reading the contents of a container whenever it's updated and giving a player an item/changing the gamemode.
